File: CS/A2404/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;
using DXSample;

namespace A2404 {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
            InitData();
        }

        #region Data initialization
        private void InitData() {
            if (null != XpoDefault.Session.FindObject<Role>(new BinaryOperator("RoleName", "Manager"))) return;
            string[] firstName = { "John", "Ben", "Mark", "William" };
            string[] lastName = { "Adams", "Clayborn", "Smith", "Watson" };
            Random rnd = new Random();
            for (int i = 0; i < 10; i++) {
                User user = new User();
                user.Name = string.Format("{0} {1}", firstName[rnd.Next(0, 4)], lastName[rnd.Next(0, 4)]);
                user.Save();
            }
            Role manager = new Role();
            manager.RoleName = "Manager";
            manager.Save();
            Role programmer = new Role();
            programmer.RoleName = "Programmer";
            programmer.Save();
            XPCollection<User> users = new XPCollection<User>();
            XPCollection<Role> roles = new XPCollection<Role>();
            foreach (User user in users) user.Roles.Add(user.Oid % 4 == 0 ? roles[0] : roles[1]);
            XpoDefault.Session.Save(users);
            XpoDefault.Session.Save(roles);
        }
        #endregion

        private void button1_Click(object sender, EventArgs e) {
            dataGridView1.DataSource = new XPCollection(typeof(User));
        }

        private void button2_Click(object sender, EventArgs e) {
            Role manager = XpoDefault.Session.FindObject<Role>(new BinaryOperator("RoleName", "Manager"));
            dataGridView1.DataSource = new XPCollection(typeof(User), new ContainsOperator("Roles", new BinaryOperator("This", manager)));
        }

        private void button3_Click(object sender, EventArgs e) {
            dataGridView1.DataSource = new XPCollection(typeof(User), new ContainsOperator("Roles", new BinaryOperator("RoleName", "Programmer")));
        }

        private void button4_Click(object sender, EventArgs e) {
            dataGridView1.DataSource = new XPCollection(typeof(User), new GroupOperator(GroupOperatorType.And, new CriteriaOperator[] {
                new ContainsOperator("Roles", new BinaryOperator("RoleName", "Programmer")),
                new BinaryOperator("Name", "%Smith%", BinaryOperatorType.Like) }));
        }
    }
}

File: VB/A2404/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering
Imports DXSample

Namespace A2404
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
			InitData()
		End Sub

		#Region "Data initialization"
		Private Sub InitData()
			If Nothing IsNot XpoDefault.Session.FindObject(Of Role)(New BinaryOperator("RoleName", "Manager")) Then
				Return
			End If
			Dim firstName() As String = { "John", "Ben", "Mark", "William" }
			Dim lastName() As String = { "Adams", "Clayborn", "Smith", "Watson" }
			Dim rnd As New Random()
			For i As Integer = 0 To 9
				Dim user As New User()
				user.Name = String.Format("{0} {1}", firstName(rnd.Next(0, 4)), lastName(rnd.Next(0, 4)))
				user.Save()
			Next i
			Dim manager As New Role()
			manager.RoleName = "Manager"
			manager.Save()
			Dim programmer As New Role()
			programmer.RoleName = "Programmer"
			programmer.Save()
			Dim users As XPCollection(Of User) = New XPCollection(Of User)()
			Dim roles As XPCollection(Of Role) = New XPCollection(Of Role)()
			For Each user As User In users
				If user.Oid Mod 4 = 0 Then
					user.Roles.Add(roles(0))
				Else
					user.Roles.Add(roles(1))
				End If
			Next user
			XpoDefault.Session.Save(users)
			XpoDefault.Session.Save(roles)
		End Sub
		#End Region

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			dataGridView1.DataSource = New XPCollection(GetType(User))
		End Sub

		Private Sub button2_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button2.Click
			Dim manager As Role = XpoDefault.Session.FindObject(Of Role)(New BinaryOperator("RoleName", "Manager"))
			dataGridView1.DataSource = New XPCollection(GetType(User), New ContainsOperator("Roles", New BinaryOperator("This", manager)))
		End Sub

		Private Sub button3_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button3.Click
			dataGridView1.DataSource = New XPCollection(GetType(User), New ContainsOperator("Roles", New BinaryOperator("RoleName", "Programmer")))
		End Sub

		Private Sub button4_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button4.Click
			dataGridView1.DataSource = New XPCollection(GetType(User), New GroupOperator(GroupOperatorType.And, New CriteriaOperator() { New ContainsOperator("Roles", New BinaryOperator("RoleName", "Programmer")), New BinaryOperator("Name", "%Smith%", BinaryOperatorType.Like) }))
		End Sub
	End Class
End Namespace